extensions/amp-form/0.1/amp-form.js

```diff
@@ -488,14 +499,48 @@ export class AmpForm {
       dev().error(TAG, 'Form submit service failed: %s', e);
     }
 
-    const varSubsFields = this.getVarSubsFields_();
-    if (this.xhrAction_) {
-      this.handleXhrSubmit_(varSubsFields, trust);
-    } else if (this.method_ == 'POST') {
-      this.handleNonXhrPost_();
-    } else if (this.method_ == 'GET') {
-      this.handleNonXhrGet_(varSubsFields);
+    // Get our special fields
+    const varSubsFields =
+      this.getVarSubsFields_();
+    const asyncInputs =
+      this.form_.getElementsByClassName(AsyncInputClasses.ASYNC_INPUT);
+
+    // Do any assertions we may need to do
+    // For NonXhrGET
+    // That way we can determine if
+    // we can submit synchronously
+    if (!this.xhrAction_ && this.method_ == 'GET') {
+      this.assertSsrTemplate_(false, 'Non-XHR GETs not supported.');
+      this.assertNoSensitiveFields_();
+
+      // If we have no async inputs, we can just submit synchronously
+      if (asyncInputs.length === 0) {
+
+        for (let i = 0; i < varSubsFields.length; i++) {
+          this.urlReplacement_.expandInputValueSync(varSubsFields[i]);
+        }
+
+        this.handleNonXhrGet_(event);
+        return Promise.resolve();
+      }
+
+      event.preventDefault();
     }
+
+    // Set ourselves to the SUBMITTING State
+    this.setState_(FormState.SUBMITTING);
+
+    // Promises to run before submitting the form
+    const presubmitPromises = [];
+    presubmitPromises.push(this.doVarSubs_(varSubsFields));
+    iterateCursor(asyncInputs, asyncInput => {
+      presubmitPromises.push(this.getValueForAsyncInput_(asyncInput));
+    });
+
+    return this.waitOnPromisesOrTimeout_(presubmitPromises, 10000).then(
+        () => this.handlePresubmitSuccess_(trust),
+        error => this.handlePresubmitError_(/**@type {!Error}*/(error))
```

Comment:
double confirming. I see `this.handlePresubmitError_` also creates a promise. So we don't need to wait for that promise here right?

Reply:
Correct. `this.submit_` used to have pieces of async within it, though it was called sync. And this error handling had promises that weren't returned as well.